quickwit-metastore/src/metastore/grpc_metastore/mod.rs

```diff
@@ -0,0 +1,754 @@
+        };
+        self.0
+            .clone()
+            .mark_splits_for_deletion(request)
+            .await
+            .map(|tonic_response| tonic_response.into_inner())
+            .map_err(|tonic_error| parse_grpc_error(&tonic_error))?;
+        Ok(())
+    }
+
+    /// Deletes a list of splits.
+    async fn delete_splits<'a>(
+        &self,
+        index_id: &str,
+        split_ids: &[&'a str],
+    ) -> MetastoreResult<()> {
+        let split_ids_vec: Vec<String> = split_ids
+            .iter()
+            .map(|split_id| split_id.to_string())
+            .collect();
+        let request = DeleteSplitsRequest {
+            index_id: index_id.to_string(),
+            split_ids: split_ids_vec,
+        };
+        self.0
+            .clone()
+            .delete_splits(request)
+            .await
+            .map(|tonic_response| tonic_response.into_inner())
+            .map_err(|tonic_error| parse_grpc_error(&tonic_error))?;
+        Ok(())
+    }
+
+    /// Adds a source to a given index.
+    async fn add_source(&self, index_id: &str, source: SourceConfig) -> MetastoreResult<()> {
+        let source_config_serialized_json =
+            serde_json::to_string(&source).map_err(|error| MetastoreError::JsonSerializeError {
+                name: "SourceConfig".to_string(),
+                message: error.to_string(),
+            })?;
+        let request = AddSourceRequest {
+            index_id: index_id.to_string(),
+            source_config_serialized_json,
+        };
+        self.0
+            .clone()
+            .add_source(request)
+            .await
+            .map(|tonic_response| tonic_response.into_inner())
+            .map_err(|tonic_error| parse_grpc_error(&tonic_error))?;
+        Ok(())
+    }
+
+    /// Removes a source from a given index.
+    async fn delete_source(&self, index_id: &str, source_id: &str) -> MetastoreResult<()> {
+        let request = DeleteSourceRequest {
+            index_id: index_id.to_string(),
+            source_id: source_id.to_string(),
+        };
+        self.0
+            .clone()
+            .delete_source(request)
+            .await
+            .map(|tonic_response| tonic_response.into_inner())
+            .map_err(|tonic_error| parse_grpc_error(&tonic_error))?;
+        Ok(())
+    }
+}
+
+fn get_metastore_grpc_addresses(members: &[ClusterMember]) -> HashSet<SocketAddr> {
+    members
+        .iter()
+        .filter(|member| {
+            member
+                .available_services
+                .contains(&QuickwitService::Metastore)
+        })
+        .map(|member| member.grpc_advertise_addr)
+        .collect()
+}
+
+/// Updates channel endpoints by:
+/// - Sending `Change::Insert` grpc addresses not already present in `grpc_addresses_in_use`.
+/// - Sending `Change::Remove` event on grpc addresses present in `grpc_addresses_in_use` but not in
+///   `updated_members`.
+async fn update_channel_endpoints(
```

Comment:
is there code we could put in common with search?

Reply:
The code is really different in the search pool. But it would be possible to put in common a diff function that would return something like a `Difference` struct with `added` and `removed` fields. I don't think it's worth it right now.